In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns 

In [2]:
# read the dataset

df = pd.read_csv('../data_source/used_cars.csv')
df.head()

brand                            model  model_year      milage  \
0      Ford  Utility Police Interceptor Base        2013  51,000 mi.   
1   Hyundai                     Palisade SEL        2021  34,742 mi.   
2     Lexus                    RX 350 RX 350        2022  22,372 mi.   
3  INFINITI                 Q50 Hybrid Sport        2015  88,900 mi.   
4      Audi        Q3 45 S line Premium Plus        2021   9,835 mi.   

       fuel_type                                             engine  \
0  E85 Flex Fuel  300.0HP 3.7L V6 Cylinder Engine Flex Fuel Capa...   
1       Gasoline                               3.8L V6 24V GDI DOHC   
2       Gasoline                                     3.5 Liter DOHC   
3         Hybrid  354.0HP 3.5L V6 Cylinder Engine Gas/Electric H...   
4       Gasoline                         2.0L I4 16V GDI DOHC Turbo   

        transmission                 ext_col int_col  \
0        6-Speed A/T                   Black   Black   
1  8-Speed Automatic         Moonlight Cloud    Gray   
2          Automatic                    Blue   Black   
3        7-Speed A/T                   Black   Black   
4  8-Speed Automatic  Glacier White Metallic   Black   

                                 accident clean_title    price  
0  At least 1 accident or damage reported         Yes  $10,300  
1  At least 1 accident or damage reported         Yes  $38,005  
2                           None reported         NaN  $54,598  
3                           None reported         Yes  $15,500  
4                           None reported         NaN  $34,999

In [5]:
df.columns

Index(['brand', 'model', 'model_year', 'milage', 'fuel_type', 'engine',
       'transmission', 'ext_col', 'int_col', 'accident', 'clean_title',
       'price'],
      dtype='object')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4009 entries, 0 to 4008
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   brand         4009 non-null   object
 1   model         4009 non-null   object
 2   model_year    4009 non-null   int64 
 3   milage        4009 non-null   object
 4   fuel_type     3839 non-null   object
 5   engine        4009 non-null   object
 6   transmission  4009 non-null   object
 7   ext_col       4009 non-null   object
 8   int_col       4009 non-null   object
 9   accident      3896 non-null   object
 10  clean_title   3413 non-null   object
 11  price         4009 non-null   object
dtypes: int64(1), object(11)
memory usage: 376.0+ KB


In [4]:
df.columns.value_counts()

brand           1
model           1
model_year      1
milage          1
fuel_type       1
engine          1
transmission    1
ext_col         1
int_col         1
accident        1
clean_title     1
price           1
Name: count, dtype: int64

In [10]:
df.shape

(4009, 12)

In [8]:
df.describe(include='all').T

count unique                         top  freq        mean  \
brand           4009     57                        Ford   386         NaN   
model           4009   1898                     M3 Base    30         NaN   
model_year    4009.0    NaN                         NaN   NaN  2015.51559   
milage          4009   2818                 110,000 mi.    16         NaN   
fuel_type       3839      7                    Gasoline  3309         NaN   
engine          4009   1146  2.0L I4 16V GDI DOHC Turbo    52         NaN   
transmission    4009     62                         A/T  1037         NaN   
ext_col         4009    319                       Black   905         NaN   
int_col         4009    156                       Black  2025         NaN   
accident        3896      2               None reported  2910         NaN   
clean_title     3413      1                         Yes  3413         NaN   
price           4009   1569                     $15,000    39         NaN   

                   std     min     25%     50%     75%     max  
brand              NaN     NaN     NaN     NaN     NaN     NaN  
model              NaN     NaN     NaN     NaN     NaN     NaN  
model_year    6.104816  1974.0  2012.0  2017.0  2020.0  2024.0  
milage             NaN     NaN     NaN     NaN     NaN     NaN  
fuel_type          NaN     NaN     NaN     NaN     NaN     NaN  
engine             NaN     NaN     NaN     NaN     NaN     NaN  
transmission       NaN     NaN     NaN     NaN     NaN     NaN  
ext_col            NaN     NaN     NaN     NaN     NaN     NaN  
int_col            NaN     NaN     NaN     NaN     NaN     NaN  
accident           NaN     NaN     NaN     NaN     NaN     NaN  
clean_title        NaN     NaN     NaN     NaN     NaN     NaN  
price              NaN     NaN     NaN     NaN     NaN     NaN

In [11]:
df.isnull().sum()

brand             0
model             0
model_year        0
milage            0
fuel_type       170
engine            0
transmission      0
ext_col           0
int_col           0
accident        113
clean_title     596
price             0
dtype: int64

In [18]:
df.columns

Index(['brand', 'model', 'model_year', 'milage', 'fuel_type', 'engine',
       'transmission', 'ext_col', 'int_col', 'accident', 'price'],
      dtype='object')

In [19]:
df['accident'].value_counts()

accident
None reported                             2910
At least 1 accident or damage reported     986
Name: count, dtype: int64

There are 170 missing values for fuel_type and 113 missing values for accident .
1. So for accident we have these unique keys and their values for accident -
     all the 113 empty values in accident has a new category now - No Information"
2. for 

In [21]:
df['accident'] = df['accident'].fillna('No information')
df['accident'].value_counts()

accident
None reported                             2910
At least 1 accident or damage reported     986
No information                             113
Name: count, dtype: int64

In [22]:
df['fuel_type'].value_counts()

fuel_type
Gasoline          3309
Hybrid             194
E85 Flex Fuel      139
Diesel             116
–                   45
Plug-In Hybrid      34
not supported        2
Name: count, dtype: int64

In [ ]:
df['fuel_type'] = df['fuel_type'].fillna('Unknown Fuel Type')
df['fuel_type'].value_counts()

fuel_type
Gasoline             3309
Hybrid                194
Unknown Fuel Type     170
E85 Flex Fuel         139
Diesel                116
–                      45
Plug-In Hybrid         34
not supported           2
Name: count, dtype: int64

In [ ]:
# after handling missing values in fuel type we unify the weird/different categories
df['fuel_type'] = df['fuel_type'].replace('–', 'Unknown Fuel Type')
df['fuel_type'] = df['fuel_type'].replace('not supported', 'Other')
df['fuel_type'].value_counts()


# ✔ 3. Show ALL rows containing a certain value
# df[df['fuel_type'] == '–']

# ✔ 4. Show random samples from the column
# df['fuel_type'].sample(10)

# ✔ 5. Show unique values with length (to detect spaces or Unicode)
# df['fuel_type'].apply(lambda x: (x, len(x))).unique()
# This helps detect hidden characters like:
# 1.trailing spaces
# 2.Unicode characters
# 3. dashes of different types

fuel_type
Gasoline             3309
Unknown Fuel Type     215
Hybrid                194
E85 Flex Fuel         139
Diesel                116
Plug-In Hybrid         34
Other                   2
Name: count, dtype: int64

Rechecking 
1. if still there are any missing values now after fixing for fuel_type & accident
2. if clean_title column has been dropped or not

In [29]:
print("dataset updated columns -----\n",df.columns)
print("\nchecking missing values after updating accident & fuel_type columns -----\n",df.isnull().sum())

dataset updated columns -----
 Index(['brand', 'model', 'model_year', 'milage', 'fuel_type', 'engine',
       'transmission', 'ext_col', 'int_col', 'accident', 'price'],
      dtype='object')

checking missing values after updating accident & fuel_type columns -----
 brand           0
model           0
model_year      0
milage          0
fuel_type       0
engine          0
transmission    0
ext_col         0
int_col         0
accident        0
price           0
dtype: int64


Text preprocessing for converting categorical(string) cols to numeric cols

In [ ]:
df['price'] = df['price'].str.replace('[\$,]', '', regex=True).astype(float)

In [32]:
df['milage'] = df['milage'].str.replace(',','').str.replace('mi.', '', regex=False).astype(float)

In [ ]:
df['engine_num'] = df['engine'].str.extract(r'(\d\.\d)').astype(float)


In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4009 entries, 0 to 4008
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   brand         4009 non-null   object 
 1   model         4009 non-null   object 
 2   model_year    4009 non-null   int64  
 3   milage        4009 non-null   float64
 4   fuel_type     4009 non-null   object 
 5   engine        4009 non-null   object 
 6   transmission  4009 non-null   object 
 7   ext_col       4009 non-null   object 
 8   int_col       4009 non-null   object 
 9   accident      4009 non-null   object 
 10  price         4009 non-null   float64
 11  engine_num    3906 non-null   float64
dtypes: float64(3), int64(1), object(8)
memory usage: 376.0+ KB


In [38]:
df['fuel_type'].unique()

array(['E85 Flex Fuel', 'Gasoline', 'Hybrid', 'Unknown Fuel Type',
       'Diesel', 'Plug-In Hybrid', '–', 'Other'], dtype=object)

In [42]:
# df['accident'].value_counts()

# Now make a binary flag:
df['accident_flag'] = df['accident'].apply(lambda x:0 if x == "None reported" else 1)

# Observation - 
1. Data is clean now with 0 missing values 
2. column clean_title - removed